discovery/stackit/stackit.go

```diff
@@ -0,0 +1,183 @@
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/version"
+
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const (
+	stackitLabelPrefix            = model.MetaLabelPrefix + "stackit_"
+	stackitLabelRole              = stackitLabelPrefix + "role"
+	stackitLabelProject           = stackitLabelPrefix + "project"
+	stackitLabelServerID          = stackitLabelPrefix + "server_id"
+	stackitLabelServerName        = stackitLabelPrefix + "server_name"
+	stackitLabelServerStatus      = stackitLabelPrefix + "server_status"
+	stackitLabelServerPowerStatus = stackitLabelPrefix + "server_power_status"
+	stackitLabelAvailabilityZone  = stackitLabelPrefix + "availability_zone"
+	stackitLabelPublicIPv4        = stackitLabelPrefix + "public_ipv4"
+)
+
+var userAgent = version.PrometheusUserAgent()
+
+// DefaultSDConfig is the default STACKIT SD configuration.
+var DefaultSDConfig = SDConfig{
+	Region:           "eu01",
+	Port:             80,
+	RefreshInterval:  model.Duration(60 * time.Second),
+	HTTPClientConfig: config.DefaultHTTPClientConfig,
+}
+
+func init() {
+	discovery.RegisterConfig(&SDConfig{})
+}
+
+// SDConfig is the configuration for STACKIT based service discovery.
+type SDConfig struct {
+	HTTPClientConfig config.HTTPClientConfig `yaml:",inline"`
+
+	RefreshInterval       model.Duration `yaml:"refresh_interval"`
+	Port                  int            `yaml:"port"`
+	Role                  Role           `yaml:"role"`
+	Region                string         `yaml:"region"`
+	Endpoint              string         `yaml:"endpoint"`
+	Project               string         `yaml:"project"`
+	ServiceAccountKey     string         `yaml:"service_account_key"`
+	PrivateKey            string         `yaml:"private_key"`
+	ServiceAccountKeyPath string         `yaml:"service_account_key_path"`
+	PrivateKeyPath        string         `yaml:"private_key_path"`
+	CredentialsFilePath   string         `yaml:"credentials_file_path"`
+
+	// For testing only
+	tokenURL string
+}
+
+// NewDiscovererMetrics implements discovery.Config.
+func (*SDConfig) NewDiscovererMetrics(_ prometheus.Registerer, rmi discovery.RefreshMetricsInstantiator) discovery.DiscovererMetrics {
+	return &stackitMetrics{
+		refreshMetrics: rmi,
+	}
+}
+
+// Name returns the name of the Config.
+func (*SDConfig) Name() string { return "stackit" }
+
+// NewDiscoverer returns a Discoverer for the Config.
+func (c *SDConfig) NewDiscoverer(opts discovery.DiscovererOptions) (discovery.Discoverer, error) {
+	return NewDiscovery(c, opts.Logger, opts.Metrics)
+}
+
+type refresher interface {
+	refresh(context.Context) ([]*targetgroup.Group, error)
+}
+
+// Role is the Role of the target within the STACKIT Ecosystem.
+type Role string
+
+// The valid options for role.
+const (
+	RoleServer Role = "server" // STACKIT IAAS API (Server)
```

Comment:
are you planning to add others any soon? if not let's just hardcode it/remove the logic and hide it from the config for now.